catroid/src/main/java/org/catrobat/catroid/web/WebConnection.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.web;
+
+import com.google.common.collect.ImmutableList;
+import com.squareup.okhttp.Call;
+import com.squareup.okhttp.Callback;
+import com.squareup.okhttp.ConnectionSpec;
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.Response;
+
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class WebConnection {
+	private final OkHttpClient okHttpClient;
+	private static final String USER_AGENT = "Mozilla/5.0 (compatible; Catrobatbot/1.0; +https://catrob.at/bot)";
+
+	private Call call;
+	private String url;
+	private WeakReference<WebRequestListener> weakListenerReference;
+
+	public WebConnection() {
+		okHttpClient = new OkHttpClient();
+		ConnectionSpec modernTlsSpec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS).build();
+		ConnectionSpec compatibleTlsSpec = new ConnectionSpec.Builder(ConnectionSpec.COMPATIBLE_TLS).build();
+		ConnectionSpec cleartextSpec = new ConnectionSpec.Builder(ConnectionSpec.CLEARTEXT).build();
+		okHttpClient.setConnectionSpecs(ImmutableList.of(modernTlsSpec, compatibleTlsSpec, cleartextSpec));
+	}
+
+	public WebConnection(OkHttpClient okHttpClient) {
+		this.okHttpClient = okHttpClient;
+	}
+
+	public void setListener(WebRequestListener listener) {
+		this.weakListenerReference = new WeakReference<>(listener);
+	}
+
+	public void setUrl(String url) {
+		this.url = url;
+	}
+
+	private synchronized WebRequestListener getListenerAndSetReferenceToNull() {
+		if (this.weakListenerReference != null) {
+			WebRequestListener listener = this.weakListenerReference.get();
+			this.weakListenerReference = null;
+			return listener;
+		}
+		return null;
+	}
+
+	public synchronized boolean sendWebRequest() {
+		try {
+			checkIfUrlIsValid();
+		} catch (RuntimeException e) {
+			return false;
```

Comment:
now that i see this. why not just call the listener with the error code and have him remove the request as usual instead of the 
boolean successfullySentWebRequest = webConnection.sendWebRequest();
if clause going on in act()?